### PR TITLE
(#1952) Fix fact filter parsing

### PR DESCRIPTION
**Pull request by ploubser**
With the addition of GJson to describe fact values the old fact parser is no longer able to correctly identify facts vs operators in examples like `'storage.#(name=="vda").size'`.

Here we rewrite the ParseFactFilterString function to no longer used complicated regular expressions. We will now correctly identify GJson facts and parse the filters correctly.

In addition we're adding a wider range of tests for ParseFactFilterString so that we're covered if we were to add more fact filter functionality in the future.

### CODE REVIEW

**Review by ripienaar — CHANGES_REQUESTED** [review]
looking pretty good, minor comments

File: filter/facts/facts_test.go [hunk 1]
```diff
@@ -133,4 +139,50 @@ var _ = Describe("Server/Discovery/Facts", func() {
 			Expect(t("fbool", "!=", "false")).To(BeTrue())
 		})
 	})
+
+	Describe("ParseFactFilterString", func() {
```
[the remark this answers is not in the document]
Comment (ripienaar) [reply filter/facts/facts_test.go:143]: Ah these https://github.com/choria-io/go-choria/blob/89da56f2d1fa15a1bc8aa8bbb096940ac32dd8fa/filter/filter_test.go#L99

Lets combine them so the function is tested one place only

File: filter/facts/facts.go [hunk 2]
```diff
@@ -173,18 +173,52 @@ func HasFact(fact string, operator string, value string, file string, log Logger
...
-		return [3]string{matched[1], "<=", matched[2]}, nil
-	} else if matched := regexp.MustCompile("^([^ ]+?)[ ]*(<=|>=|<|>|!=|==|=~)[ ]*(.+)").FindStringSubmatch(f); len(matched) > 0 {
-		return [3]string{matched[1], matched[2], matched[3]}, nil
-	} else if matched := regexp.MustCompile("^(.+?)[ ]*=[ ]*/(.+)/$").FindStringSubmatch(f); len(matched) > 0 {
-		return [3]string{matched[1], "=~", "/" + matched[2] + "/"}, nil
-	} else if matched := regexp.MustCompile("^([^= ]+?)[ ]*=[ ]*(.+)").FindStringSubmatch(f); len(matched) > 0 {
-		return [3]string{matched[1], "==", matched[2]}, nil
+func ParseFactFilterString(f string) ([3]string, error) {
+	// strip whatspaces before processing
+	f = strings.ReplaceAll(f, " ", "")
```
[the remark this answers is not in the document]
Comment (ploubser) [reply filter/facts/facts.go:178]: So this happens before we identify any of the 3 tokens that make up a valid fact filter. I agree that using Trimspace() is a more elegant solution to removing the whitespace, but we would have to call it on both the left and right operands later in the function. Trimspace also might fall apart if this is a valid fact filter

```
storage.#(name == "vda").size = 64
```

File: filter/facts/facts.go [hunk 3]
```diff
@@ -173,18 +173,52 @@ func HasFact(fact string, operator string, value string, file string, log Logger
...
+	validOperators := regexp.MustCompile(`<=|>=|=>|=<|<|>|!=|=~|={1,2}`)
```
[the remark this answers is not in the document]
Comment (ploubser) [reply filter/facts/facts.go:179]: 👍

File: filter/facts/facts_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ploubser) [reply filter/facts/facts_test.go:143]: 👍

File: filter/facts/facts.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ripienaar) [reply filter/facts/facts.go:178]: Ah, I see, but we also cant just remove all the whitespace though can we? What If I want to match against `"Foo Bar"` wont that turn it into `"FooBar"`?

[the remark this answers is not in the document]
Comment (ploubser) [reply filter/facts/facts.go:178]: Yep, that's correct. Let me rethink this.

File: filter/filter_test.go [hunk 4]
```diff
@@ -160,10 +161,12 @@ var _ = Describe("Filter", func() {
 			t("foo == bar", "foo", "==", "bar")
 		})
 
-		It("Should fail for facts in the wrong format", func() {
-			pf, err := ParseFactFilterString("foo")
-			Expect(err).To(MatchError("could not parse fact foo it does not appear to be in a valid format"))
-			Expect(pf).To(BeNil())
+		It("Should fail on invalid fact filters", func() {
```
Comment (ripienaar) [inline filter/filter_test.go:164]: looks great, lets add a few tests against these new kinds of filters that would have triggered the bug and this one is done!

File: filter/filter_test.go [hunk 5]
```diff
@@ -160,10 +161,20 @@ var _ = Describe("Filter", func() {
...
-		It("Should fail for facts in the wrong format", func() {
-			pf, err := ParseFactFilterString("foo")
-			Expect(err).To(MatchError("could not parse fact foo it does not appear to be in a valid format"))
-			Expect(pf).To(BeNil())
+		It("Should parse gjson facts", func() {
+			t("storage.#(name=\"nvme0n1\").size==64", "storage.#(name=\"nvme0n1\").size", "==", "64")
+			t("storage.#(name==\"nvme0n1\").size=64", "storage.#(name==\"nvme0n1\").size", "==", "64")
+			t("storage.#(name=\"nvme0n1\").size == 64", "storage.#(name=\"nvme0n1\").size", "==", "64")
+			t("storage.#(name<= \"nvme0n1\" ).size==64", "storage.#(name<= \"nvme0n1\" ).size", "==", "64")
+			t("storage.#(name=\"foo bar\").size=>baz bar", "storage.#(name=\"foo bar\").size", ">=", "baz bar")
```
Comment (ripienaar) [inline filter/filter_test.go:169]: Also works with single quotes like `storage.#(name=\"nvme0n1\").size=64`?

last one I think :)

**Review by ripienaar — APPROVED** [review]
LGTM!

**Pull request merged**